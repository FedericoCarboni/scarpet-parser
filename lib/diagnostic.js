/**
 * Position in a source document expressed as zero-based UTF-16 character offsets.
 *
 * This definition is compatible with the Language Server Protocol, but adds `offset`.
 *
 * @typedef {object} Position
 * @property {number} offset Character offset in the source code (zero-based).
 * @property {number} line Line position in the source code (zero-based).
 * @property {number} character Character offset on a line in the source code (zero-based).
 */

/**
 * Range in a source document.
 *
 * @typedef {object} Range
 * @property {Position} start
 * @property {Position} end
 */

/**
 * A diagnostic message compatible with [LSP `Diagnostic`][1].
 *
 * [1]: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnostic
 */
export class Diagnostic {
    /**
     * @param {DiagnosticSeverity} severity
     * @param {Range} range
     * @param {DiagnosticCode} code
     * @internal
     */
    constructor(severity, range, code) {
        this.range = {start: range.start, end: range.end};
        this.severity = severity;
        this.code = code;
        this.source = 'scarpet';
    }

    /** Get a friendly message describing the problem. */
    get message() {
        return DIAGNOSTIC_MESSAGES[this.code];
    }
}

let i = 0;
/**
 * Enumeration of possible diagnostics.
 *
 * @enum {number}
 */
export const DiagnosticCode = Object.freeze({
    // General
    UnexpectedEof: i++,
    UnexpectedToken: i++,
    UnknownOperator: i++,
    // Number literals
    ExpectedHexDigit: i++,
    ExpectedDecimalDigit: i++,
    LongLossOfPrecision: i++,
    DoubleLossOfPrecision: i++,
    MoreThanOnePoint: i++,
    NoExponentDigits: i++,
    UnicodeDigit: i++,
    // String literal
    UnknownEscapeSequence: i++,
    // Tokenizer options
    UnexpectedNewLineMarker: i++,
    UnexpectedComment: i++,
    // Functions
    MismatchedCloseParen: i++,
    UnmatchedCloseParen: i++,
    EmptyExpression: i++,
    MissingSemicolon: i++,
    InvalidArrowOperator: i++,
    InvalidFunctionSignature: i++,
    RestParameterAlreadyDefined: i++,
    OuterTakesOneVariable: i++,
    AssignToConstant: i++,
    CannotAssign: i++,
    AssignToSystemVariable: i++,
});

const DIAGNOSTIC_MESSAGES = Object.freeze([
    'Unexpected end of file (is it truncated?)',
    'Unexpected token',
    'Unknown operator',
    'Expected hexadecimal digit after 0x (did you mean 0x0?)',
    'Expected decimal digit',
    'Number literals greater than 2^63-1 (or less than -2^63) lose precision at runtime',
    'Number literal loses precision at runtime',
    'Number literal contains more than one decimal point',
    'Number literal exponent has no digits',
    'Number literal contains non ASCII digit',
    "Invalid escape sequence (only `\\n`, `\\t`, `\\'` and `\\\\` are supported)",
    'Unexpected new line marker $ (not allowed in script files)',
    'Unexpected comment // (not allowed in /script run)',
    'Mismatched closing parenthesis, using parentheses like (} is confusing',
    'Unmatched closing parenthesis',
    'Empty expressions are not allowed, specify a null to suppress the error',
    'Missing semicolon',
    'Expected a function signature before the ->',
    'Function signatures can only contain parameters or rest parameters',
    'Variable argument (...args) parameter is already defined',
    'Outer expects a single variable',
    'Cannot assign a new value to a constant',
    'Can only assign to variables and attributes',
    'System variables _, _i, _a, _x, _y, _z and _trace are readonly',
]);

/** @enum {1 | 2 | 3 | 4} */
export const DiagnosticSeverity = Object.freeze({
    Error: 1,
    Warning: 2,
    Information: 3,
    Hint: 4,
});
