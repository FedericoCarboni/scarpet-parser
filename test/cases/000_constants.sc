// 
euler;pi;null;true;false