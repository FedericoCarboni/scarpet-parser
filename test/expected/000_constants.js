import {Diagnostic} from '../../lib/diagnostic.js';

/** @type {Diagnostic[]} */
export const errors = [];

/** @type {import('../../lib/Parser.js').Node} */
export const root = {
    kind: 'BinaryExpression',
    operator: ';',
    lvalue: {
        kind: 'BinaryExpression',
        operator: ';',
        lvalue: {
            kind: 'BinaryExpression',
            operator: ';',
            lvalue: {
                kind: 'BinaryExpression',
                operator: ';',
                lvalue: {
                    kind: 'Constant',
                    name: 'euler',
                    start: {
                        offset: 4,
                        line: 1,
                        character: 0,
                    },
                    end: {
                        offset: 9,
                        line: 1,
                        character: 5,
                    },
                },
                rvalue: {
                    kind: 'Constant',
                    name: 'pi',
                    start: {
                        offset: 10,
                        line: 1,
                        character: 6,
                    },
                    end: {
                        offset: 12,
                        line: 1,
                        character: 8,
                    },
                },
                comment: '// \n',
                start: {
                    offset: 4,
                    line: 1,
                    character: 0,
                },
                end: {
                    offset: 12,
                    line: 1,
                    character: 8,
                },
            },
            rvalue: {
                kind: 'Constant',
                name: 'null',
                start: {
                    offset: 13,
                    line: 1,
                    character: 9,
                },
                end: {
                    offset: 17,
                    line: 1,
                    character: 13,
                },
            },
            comment: '// \n',
            start: {
                offset: 4,
                line: 1,
                character: 0,
            },
            end: {
                offset: 17,
                line: 1,
                character: 13,
            },
        },
        rvalue: {
            kind: 'Constant',
            name: 'true',
            start: {
                offset: 18,
                line: 1,
                character: 14,
            },
            end: {
                offset: 22,
                line: 1,
                character: 18,
            },
        },
        comment: '// \n',
        start: {
            offset: 4,
            line: 1,
            character: 0,
        },
        end: {
            offset: 22,
            line: 1,
            character: 18,
        },
    },
    rvalue: {
        kind: 'Constant',
        name: 'false',
        start: {
            offset: 23,
            line: 1,
            character: 19,
        },
        end: {
            offset: 28,
            line: 1,
            character: 24,
        },
    },
    comment: '// \n',
    start: {
        offset: 4,
        line: 1,
        character: 0,
    },
    end: {
        offset: 28,
        line: 1,
        character: 24,
    },
};
