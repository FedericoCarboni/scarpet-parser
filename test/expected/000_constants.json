{
  "errors": [],
  "warnings": [],
  "root": {
    "kind": "BinaryExpression",
    "operator": ";",
    "lvalue": {
      "kind": "BinaryExpression",
      "operator": ";",
      "lvalue": {
        "kind": "BinaryExpression",
        "operator": ";",
        "lvalue": {
          "kind": "BinaryExpression",
          "operator": ";",
          "lvalue": {
            "kind": "Constant",
            "name": "euler",
            "start": {
              "pos": 4,
              "row": 1,
              "col": 0
            },
            "end": {
              "pos": 9,
              "row": 1,
              "col": 5
            }
          },
          "rvalue": {
            "kind": "Constant",
            "name": "pi",
            "start": {
              "pos": 10,
              "row": 1,
              "col": 6
            },
            "end": {
              "pos": 12,
              "row": 1,
              "col": 8
            }
          },
          "comment": "// \n",
          "start": {
            "pos": 4,
            "row": 1,
            "col": 0
          },
          "end": {
            "pos": 12,
            "row": 1,
            "col": 8
          }
        },
        "rvalue": {
          "kind": "Constant",
          "name": "null",
          "start": {
            "pos": 13,
            "row": 1,
            "col": 9
          },
          "end": {
            "pos": 17,
            "row": 1,
            "col": 13
          }
        },
        "comment": "// \n",
        "start": {
          "pos": 4,
          "row": 1,
          "col": 0
        },
        "end": {
          "pos": 17,
          "row": 1,
          "col": 13
        }
      },
      "rvalue": {
        "kind": "Constant",
        "name": "true",
        "start": {
          "pos": 18,
          "row": 1,
          "col": 14
        },
        "end": {
          "pos": 22,
          "row": 1,
          "col": 18
        }
      },
      "comment": "// \n",
      "start": {
        "pos": 4,
        "row": 1,
        "col": 0
      },
      "end": {
        "pos": 22,
        "row": 1,
        "col": 18
      }
    },
    "rvalue": {
      "kind": "Constant",
      "name": "false",
      "start": {
        "pos": 23,
        "row": 1,
        "col": 19
      },
      "end": {
        "pos": 28,
        "row": 1,
        "col": 24
      }
    },
    "comment": "// \n",
    "start": {
      "pos": 4,
      "row": 1,
      "col": 0
    },
    "end": {
      "pos": 28,
      "row": 1,
      "col": 24
    }
  }
}